### PR TITLE
Adapt openldap-pbkfd2 module to use MozNSS

### PR DESCRIPTION
**Pull request by azmagillian**
Hi @hamano and community plz review and comment my proposed contribution.
The applied changes introduce use of MozNSS library in order for an OpenLDAP configured with moznss option to be able to use openldap-pbkfd2 module!

### CODE REVIEW

**Comment by azmagillian** [comment]
One comment! I had to change my spec file in order to "make"  the module after my changes. To be specific i had to add INCS+="-I../../../../include/nspr -I../../../../include -I../../../../servers/slapd".
Should i commit this change also in the Makefile?

**Comment by hamano** [comment]
Thanks for your contribution!
I'll review your code in a week.

File: pw-pbkdf2.c [hunk 1]
```diff
@@ -229,6 +246,95 @@ static int pbkdf2_encrypt(
...
 						  dk.bv_len, dk_value);
+#elif HAVE_MOZNSS
+	SECStatus rv;
+	rv = NSS_NoDB_Init(".");
+	if (rv != SECSuccess)
+ 	{
+    		fprintf(stderr, "NSS initialization failed (err %d)\n",
+        	PR_GetError());
+  	}
+        PK11SlotInfo *slot = NULL;
```
Comment (hamano) [inline pw-pbkdf2.c]: Some Indentations are wrong, tab indentation are used in upstream OpenLDAP project.

File: pw-pbkdf2.c [hunk 2]
```diff
@@ -229,6 +246,95 @@ static int pbkdf2_encrypt(
...
+	SECAlgorithmID * algid = NULL;
+	PRArenaPool *arena1;
+	PRArenaPool *arena2;
+	PRArenaPool *arena3;
+	arena1 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+	arena2 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+	arena3 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
```
Comment (hamano) [inline pw-pbkdf2.c]: Please lease perform suitable naming. ex: xxx_arena, arena_xxx.
BTW,, are these three independent arenas required?

File: pw-pbkdf2.c [hunk 3]
```diff
@@ -229,6 +246,95 @@ static int pbkdf2_encrypt(
...
+                algid=PK11_CreatePBEV2AlgorithmID(SEC_OID_PKCS5_PBKDF2, SEC_OID_HMAC_SHA1, SEC_OID_HMAC_SHA1, PBKDF2_SHA1_DK_SIZE, iteration, &salt_moz);
+                PK11SymKey *key = NULL;
+                key=PK11_PBEKeyGen(slot, algid, &pwitem, 0, NULL);
+                SECStatus rv2 = PK11_ExtractKeyValue(key);
+                SECItem *symkey;
+                SECITEM_AllocItem(arena3, symkey, PBKDF2_SHA1_DK_SIZE);
+                symkey=PK11_GetKeyData(key);
+                dk.bv_val=memcpy(dk.bv_val, symkey->data, symkey->len);
+                PK11_FreeSymKey(key);
+        }else if(!ber_bvcmp(scheme, &pbkdf2_sha1_scheme)){
```
Comment (hamano) [inline pw-pbkdf2.c]: You should write:
~~~
if(!ber_bvcmp(scheme, &pbkdf2_scheme) || !ber_bvcmp(scheme, &pbkdf2_sha1_scheme)){
~~~
We should write the same in OpenSSL GnuTLS :p

File: pw-pbkdf2.c (hunk 1 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: ok i retabulated the whole file

File: pw-pbkdf2.c (hunk 3 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: done!

Comment (azmagillian) [reply pw-pbkdf2.c]: i suppose this should be applied in all checks of ber_bvcmp(scheme, &pbkdf2_sha1_scheme)

File: pw-pbkdf2.c (hunk 2 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: ok about the naming..not sure if they are all needed

File: pw-pbkdf2.c (hunk 3 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: did it for GnuTLS also

**Comment by azmagillian** [comment]
I wiil test it locally to verify i didn't mess with the functionality!
In the meantime plz continue your review !

File: pw-pbkdf2.c [hunk 4]
```diff
@@ -229,6 +235,81 @@ static int pbkdf2_encrypt(
...
+	SECStatus rv;
+	rv = NSS_NoDB_Init(".");
+	if (rv != SECSuccess)
+	{
+		fprintf(stderr, "NSS initialization failed (err %d)\n",
+		PR_GetError());
+	}
+	PK11SlotInfo *slot = NULL;
+	SECAlgorithmID * algid = NULL;
+	PRArenaPool *arena_001;
```
Comment (hamano) [inline pw-pbkdf2.c]: for example: salt_arena

File: pw-pbkdf2.c [hunk 5]
```diff
@@ -229,6 +235,81 @@ static int pbkdf2_encrypt(
 						  dk.bv_len, iteration,
 						  salt.bv_len, (const uint8_t *) salt.bv_val,
 						  dk.bv_len, dk_value);
+#elif HAVE_MOZNSS
+	SECStatus rv;
+	rv = NSS_NoDB_Init(".");
+	if (rv != SECSuccess)
+	{
+		fprintf(stderr, "NSS initialization failed (err %d)\n",
+		PR_GetError());
```
Comment (hamano) [inline pw-pbkdf2.c]: return LUTIL_PASSWD_ERR;

File: pw-pbkdf2.c [hunk 6]
```diff
@@ -229,6 +235,81 @@ static int pbkdf2_encrypt(
...
+	PRArenaPool *arena_003;
+	arena_001 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+	arena_002 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+	arena_003 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+	SECItem salt_moz={ siBuffer, NULL, 0};
+	SECItem pwitem={ siBuffer, NULL, 0};
+	slot=PK11_GetBestSlot(SEC_OID_PKCS5_PBKDF2, NULL);
+	if (slot == NULL)
+	{
+		printf("error initializing slot");
```
Comment (hamano) [inline pw-pbkdf2.c]: output to use stderr and
return LUTIL_PASSWD_ERR;

File: pw-pbkdf2.c [hunk 7]
```diff
@@ -229,6 +235,81 @@ static int pbkdf2_encrypt(
...
+	{
+		fprintf(stderr, "NSS initialization failed (err %d)\n",
+		PR_GetError());
+	}
+	PK11SlotInfo *slot = NULL;
+	SECAlgorithmID * algid = NULL;
+	PRArenaPool *arena_001;
+	PRArenaPool *arena_002;
+	PRArenaPool *arena_003;
+	arena_001 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
```
Comment (hamano) [inline pw-pbkdf2.c]: We must assume Insufficient memory.  NULL check.

File: pw-pbkdf2.c [hunk 8]
```diff
@@ -399,15 +481,95 @@ static int pbkdf2_check(
 						  dk_len, iteration,
 						  PBKDF2_SALT_SIZE, salt_value,
 						  dk_len, input_dk_value);
+
+#elif HAVE_MOZNSS
+	SECStatus rv;
+	rv = NSS_NoDB_Init(".");
```
Comment (hamano) [inline pw-pbkdf2.c]: https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/SSL_functions/sslfnc.html#1234224
> NSS_NoDB_Init isnot idempotent, so call it only once.

so, move to init_module().

File: pw-pbkdf2.c [hunk 9]
```diff
@@ -399,15 +481,95 @@ static int pbkdf2_check(
...
+	if (rv != SECSuccess)
+	{
+		fprintf(stderr, "NSS initialization failed (err %d)\n",
+		PR_GetError());
```
Comment (hamano) [inline pw-pbkdf2.c]: return LUTIL_PASSWD_ERR;

File: pw-pbkdf2.c [hunk 10]
```diff
@@ -229,6 +235,81 @@ static int pbkdf2_encrypt(
...
+	SECItem salt_moz={ siBuffer, NULL, 0};
+	SECItem pwitem={ siBuffer, NULL, 0};
+	slot=PK11_GetBestSlot(SEC_OID_PKCS5_PBKDF2, NULL);
+	if (slot == NULL)
+	{
+		printf("error initializing slot");
+	}
+
+	if(!ber_bvcmp(scheme, &pbkdf2_scheme) || !ber_bvcmp(scheme, &pbkdf2_sha1_scheme)){
+		SECITEM_AllocItem(arena_001, &salt_moz, (unsigned int) PBKDF2_SALT_SIZE);
```
Comment (hamano) [inline pw-pbkdf2.c]: NULL check SECITEM_AllocItem.

File: pw-pbkdf2.c [hunk 11]
```diff
@@ -229,6 +235,81 @@ static int pbkdf2_encrypt(
...
+		salt_moz.data = memcpy(salt_moz.data , salt_value, PBKDF2_SALT_SIZE);
+		SECITEM_AllocItem(arena_002, &pwitem, (unsigned int)passwd->bv_len);
+		pwitem.data= memcpy(pwitem.data, passwd->bv_val, passwd->bv_len );
+		algid=PK11_CreatePBEV2AlgorithmID(SEC_OID_PKCS5_PBKDF2, SEC_OID_HMAC_SHA1, SEC_OID_HMAC_SHA1, PBKDF2_SHA1_DK_SIZE, iteration, &salt_moz);
+		PK11SymKey *key = NULL;
+		key=PK11_PBEKeyGen(slot, algid, &pwitem, 0, NULL);
+		SECStatus rv2 = PK11_ExtractKeyValue(key);
+		SECItem *symkey;
+		SECITEM_AllocItem(arena_003, symkey, PBKDF2_SHA1_DK_SIZE);
+		symkey=PK11_GetKeyData(key);
```
Comment (hamano) [inline pw-pbkdf2.c]: NULL check SECITEM_AllocItem.

File: pw-pbkdf2.c [hunk 12]
```diff
@@ -399,15 +481,95 @@ static int pbkdf2_check(
...
+	PRArenaPool *arena_003;
+	arena_001 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+	arena_002 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+	arena_003 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+	SECItem salt_moz={ siBuffer, NULL, 0};
+	SECItem pwitem={ siBuffer, NULL, 0};
+	slot=PK11_GetBestSlot(SEC_OID_PKCS5_PBKDF2, NULL);
+	if (slot == NULL)
+	{
+		printf("error initializing slot");
```
Comment (hamano) [inline pw-pbkdf2.c]: return LUTIL_PASSWD_ERR;

File: pw-pbkdf2.c [hunk 13]
```diff
@@ -399,15 +481,95 @@ static int pbkdf2_check(
...
+		PK11_FreeSymKey(key);
+	}else{
+		return LUTIL_PASSWD_ERR;
+	}
+
+
+		PK11_FreeSlot(slot);
+		PORT_FreeArena(arena_001, PR_FALSE);
+		PORT_FreeArena(arena_002, PR_FALSE);
+		PORT_FreeArena(arena_003, PR_FALSE);
```
Comment (hamano) [inline pw-pbkdf2.c]: indent

**Comment by hamano** [comment]
NULL check is very important for authentication system, because it may become security hole.

File: pw-pbkdf2.c (hunk 5 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: Done

File: pw-pbkdf2.c (hunk 9 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: Done

File: pw-pbkdf2.c (hunk 13 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: Done

File: pw-pbkdf2.c (hunk 12 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: Done

File: pw-pbkdf2.c (hunk 11 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: Done

File: pw-pbkdf2.c (hunk 10 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: Done

File: pw-pbkdf2.c (hunk 8 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: Done

File: pw-pbkdf2.c (hunk 7 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: Done

File: pw-pbkdf2.c (hunk 6 above)
Comment (azmagillian) [reply pw-pbkdf2.c]: Done

**Comment by azmagillian** [comment]
Verified functionality too.

**Pull request merged**